File: py/railwaystationphotos_sdk.py
# RailwayStationPhotos SDK

from utility.voxgig_struct import voxgig_struct as vs
from core.utility_type import RailwayStationPhotosUtility
from core.spec import RailwayStationPhotosSpec
from core import helpers

# Load utility registration (populates Utility._registrar)
from utility import register

# Load features
from feature.base_feature import RailwayStationPhotosBaseFeature
from features import _make_feature


class RailwayStationPhotosSDK:

    def __init__(self, options=None):
        self.mode = "live"
        self.features = []
        self.options = None

        utility = RailwayStationPhotosUtility()
        self._utility = utility

        from config import make_config
        config = make_config()

        self._rootctx = utility.make_context({
            "client": self,
            "utility": utility,
            "config": config,
            "options": options if options is not None else {},
            "shared": {},
        }, None)

        self.options = utility.make_options(self._rootctx)

        if vs.getpath(self.options, "feature.test.active") is True:
            self.mode = "test"

        self._rootctx.options = self.options

        # Add features from config.
        feature_opts = helpers.to_map(vs.getprop(self.options, "feature"))
        if feature_opts is not None:
            feature_items = vs.items(feature_opts)
            if feature_items is not None:
                for item in feature_items:
                    fname = item[0]
                    fopts = helpers.to_map(item[1])
                    if fopts is not None and fopts.get("active") is True:
                        utility.feature_add(self._rootctx, _make_feature(fname))

        # Add extension features.
        extend = vs.getprop(self.options, "extend")
        if isinstance(extend, list):
            for f in extend:
                if isinstance(f, dict) or (hasattr(f, "get_name") and callable(f.get_name)):
                    utility.feature_add(self._rootctx, f)

        # Initialize features.
        for f in self.features:
            utility.feature_init(self._rootctx, f)

        utility.feature_hook(self._rootctx, "PostConstruct")

        # #BuildFeatures

    def options_map(self):
        out = vs.clone(self.options)
        if isinstance(out, dict):
            return out
        return {}

    def get_utility(self):
        return RailwayStationPhotosUtility.copy(self._utility)

    def get_root_ctx(self):
        return self._rootctx

    def prepare(self, fetchargs=None):
        utility = self._utility

        if fetchargs is None:
            fetchargs = {}

        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "prepare",
            "ctrl": ctrl,
        }, self._rootctx)

        options = self.options

        path = vs.getprop(fetchargs, "path") or ""
        if not isinstance(path, str):
            path = ""

        method = vs.getprop(fetchargs, "method") or "GET"
        if not isinstance(method, str):
            method = "GET"

        params = helpers.to_map(vs.getprop(fetchargs, "params"))
        if params is None:
            params = {}
        query = helpers.to_map(vs.getprop(fetchargs, "query"))
        if query is None:
            query = {}

        headers = utility.prepare_headers(ctx)

        base = vs.getprop(options, "base") or ""
        if not isinstance(base, str):
            base = ""
        prefix = vs.getprop(options, "prefix") or ""
        if not isinstance(prefix, str):
            prefix = ""
        suffix = vs.getprop(options, "suffix") or ""
        if not isinstance(suffix, str):
            suffix = ""

        ctx.spec = RailwayStationPhotosSpec({
            "base": base,
            "prefix": prefix,
            "suffix": suffix,
            "path": path,
            "method": method,
            "params": params,
            "query": query,
            "headers": headers,
            "body": vs.getprop(fetchargs, "body"),
            "step": "start",
        })

        # Merge user-provided headers.
        uh = vs.getprop(fetchargs, "headers")
        if isinstance(uh, dict):
            for k, v in uh.items():
                ctx.spec.headers[k] = v

        _, err = utility.prepare_auth(ctx)
        if err is not None:
            raise err

        fetchdef, err = utility.make_fetch_def(ctx)
        if err is not None:
            raise err

        return fetchdef

    def direct(self, fetchargs=None):
        utility = self._utility

        try:
            fetchdef = self.prepare(fetchargs)
        except Exception as err:
            # direct() is the raw-HTTP escape hatch: it never raises, it
            # returns a result object callers branch on via result["ok"].
            return {"ok": False, "err": err}

        if fetchargs is None:
            fetchargs = {}
        ctrl = helpers.to_map(vs.getprop(fetchargs, "ctrl"))
        if ctrl is None:
            ctrl = {}

        ctx = utility.make_context({
            "opname": "direct",
            "ctrl": ctrl,
        }, self._rootctx)

        url = fetchdef.get("url", "")
        fetched, fetch_err = utility.fetcher(ctx, url, fetchdef)

        if fetch_err is not None:
            return {"ok": False, "err": fetch_err}

        if fetched is None:
            return {
                "ok": False,
                "err": ctx.make_error("direct_no_response", "response: undefined"),
            }

        if isinstance(fetched, dict):
            status = helpers.to_int(vs.getprop(fetched, "status"))
            headers = vs.getprop(fetched, "headers") or {}

            # No-body responses (204, 304) and explicit zero content-length
            # must skip JSON parsing — calling json() on an empty body raises.
            content_length = None
            if isinstance(headers, dict):
                content_length = headers.get("content-length")
            no_body = status in (204, 304) or str(content_length) == "0"

            json_data = None
            if not no_body:
                jf = vs.getprop(fetched, "json")
                if callable(jf):
                    try:
                        json_data = jf()
                    except Exception:
                        # Non-JSON body (e.g. text/plain, text/html). Surface
                        # status + headers but leave data as None.
                        json_data = None

            return {
                "ok": status >= 200 and status < 300,
                "status": status,
                "headers": headers,
                "data": json_data,
            }

        return {
            "ok": False,
            "err": ctx.make_error("direct_invalid", "invalid response type"),
        }


    def AdminInbox(self, data=None) -> "AdminInboxEntity":
        """Entity factory: client.AdminInbox().list() / client.AdminInbox().load({"id": ...})."""
        from entity.admin_inbox_entity import AdminInboxEntity
        return AdminInboxEntity(self, data)


    def Country(self, data=None) -> "CountryEntity":
        """Entity factory: client.Country().list() / client.Country().load({"id": ...})."""
        from entity.country_entity import CountryEntity
        return CountryEntity(self, data)


    def Inbox(self, data=None) -> "InboxEntity":
        """Entity factory: client.Inbox().list() / client.Inbox().load({"id": ...})."""
        from entity.inbox_entity import InboxEntity
        return InboxEntity(self, data)


    def InboxCount(self, data=None) -> "InboxCountEntity":
        """Entity factory: client.InboxCount().list() / client.InboxCount().load({"id": ...})."""
        from entity.inbox_count_entity import InboxCountEntity
        return InboxCountEntity(self, data)


    def InboxEntry(self, data=None) -> "InboxEntryEntity":
        """Entity factory: client.InboxEntry().list() / client.InboxEntry().load({"id": ...})."""
        from entity.inbox_entry_entity import InboxEntryEntity
        return InboxEntryEntity(self, data)


    def InboxStateQuery(self, data=None) -> "InboxStateQueryEntity":
        """Entity factory: client.InboxStateQuery().list() / client.InboxStateQuery().load({"id": ...})."""
        from entity.inbox_state_query_entity import InboxStateQueryEntity
        return InboxStateQueryEntity(self, data)


    def OAuthToken(self, data=None) -> "OAuthTokenEntity":
        """Entity factory: client.OAuthToken().list() / client.OAuthToken().load({"id": ...})."""
        from entity.o_auth_token_entity import OAuthTokenEntity
        return OAuthTokenEntity(self, data)


    def Oauth(self, data=None) -> "OauthEntity":
        """Entity factory: client.Oauth().list() / client.Oauth().load({"id": ...})."""
        from entity.oauth_entity import OauthEntity
        return OauthEntity(self, data)


    def Photo(self, data=None) -> "PhotoEntity":
        """Entity factory: client.Photo().list() / client.Photo().load({"id": ...})."""
        from entity.photo_entity import PhotoEntity
        return PhotoEntity(self, data)


    def PhotoDownload(self, data=None) -> "PhotoDownloadEntity":
        """Entity factory: client.PhotoDownload().list() / client.PhotoDownload().load({"id": ...})."""
        from entity.photo_download_entity import PhotoDownloadEntity
        return PhotoDownloadEntity(self, data)


    def PhotoStation(self, data=None) -> "PhotoStationEntity":
        """Entity factory: client.PhotoStation().list() / client.PhotoStation().load({"id": ...})."""
        from entity.photo_station_entity import PhotoStationEntity
        return PhotoStationEntity(self, data)


    def PhotoUpload(self, data=None) -> "PhotoUploadEntity":
        """Entity factory: client.PhotoUpload().list() / client.PhotoUpload().load({"id": ...})."""
        from entity.photo_upload_entity import PhotoUploadEntity
        return PhotoUploadEntity(self, data)


    def Photographer(self, data=None) -> "PhotographerEntity":
        """Entity factory: client.Photographer().list() / client.Photographer().load({"id": ...})."""
        from entity.photographer_entity import PhotographerEntity
        return PhotographerEntity(self, data)


    def Profile(self, data=None) -> "ProfileEntity":
        """Entity factory: client.Profile().list() / client.Profile().load({"id": ...})."""
        from entity.profile_entity import ProfileEntity
        return ProfileEntity(self, data)


    def PublicInbox(self, data=None) -> "PublicInboxEntity":
        """Entity factory: client.PublicInbox().list() / client.PublicInbox().load({"id": ...})."""
        from entity.public_inbox_entity import PublicInboxEntity
        return PublicInboxEntity(self, data)


    def Stat(self, data=None) -> "StatEntity":
        """Entity factory: client.Stat().list() / client.Stat().load({"id": ...})."""
        from entity.stat_entity import StatEntity
        return StatEntity(self, data)



    @classmethod
    def test(cls, testopts=None, sdkopts=None) -> "RailwayStationPhotosSDK":
        if sdkopts is None:
            sdkopts = {}
        sdkopts = vs.clone(sdkopts)
        if not isinstance(sdkopts, dict):
            sdkopts = {}

        if testopts is None:
            testopts = {}
        testopts = vs.clone(testopts)
        if not isinstance(testopts, dict):
            testopts = {}
        testopts["active"] = True

        vs.setpath(sdkopts, "feature.test", testopts)

        sdk = cls(sdkopts)
        sdk.mode = "test"

        return sdk


from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from entity.admin_inbox_entity import AdminInboxEntity
    from entity.country_entity import CountryEntity
    from entity.inbox_entity import InboxEntity
    from entity.inbox_count_entity import InboxCountEntity
    from entity.inbox_entry_entity import InboxEntryEntity
    from entity.inbox_state_query_entity import InboxStateQueryEntity
    from entity.o_auth_token_entity import OAuthTokenEntity
    from entity.oauth_entity import OauthEntity
    from entity.photo_entity import PhotoEntity
    from entity.photo_download_entity import PhotoDownloadEntity
    from entity.photo_station_entity import PhotoStationEntity
    from entity.photo_upload_entity import PhotoUploadEntity
    from entity.photographer_entity import PhotographerEntity
    from entity.profile_entity import ProfileEntity
    from entity.public_inbox_entity import PublicInboxEntity
    from entity.stat_entity import StatEntity

File: php/railwaystationphotos_sdk.php
<?php
declare(strict_types=1);

// RailwayStationPhotos SDK

require_once __DIR__ . '/utility/struct/Struct.php';
require_once __DIR__ . '/core/UtilityType.php';
require_once __DIR__ . '/core/Spec.php';
require_once __DIR__ . '/core/Helpers.php';

// Load utility registration
require_once __DIR__ . '/utility/Register.php';

// Load config and features
require_once __DIR__ . '/config.php';
require_once __DIR__ . '/feature/BaseFeature.php';
require_once __DIR__ . '/features.php';

use Voxgig\Struct\Struct;

class RailwayStationPhotosSDK
{
    public string $mode;
    public array $features;
    public ?array $options;

    private $_utility;
    private $_rootctx;

    public function __construct(array $options = [])
    {
        $this->mode = "live";
        $this->features = [];
        $this->options = null;

        $utility = new RailwayStationPhotosUtility();
        $this->_utility = $utility;

        $config = RailwayStationPhotosConfig::make_config();

        $this->_rootctx = ($utility->make_context)([
            "client" => $this,
            "utility" => $utility,
            "config" => $config,
            "options" => $options ?? [],
            "shared" => [],
        ], null);

        $this->options = ($utility->make_options)($this->_rootctx);

        if (Struct::getpath($this->options, "feature.test.active") === true) {
            $this->mode = "test";
        }

        $this->_rootctx->options = $this->options;

        // Add features from config.
        $feature_opts = RailwayStationPhotosHelpers::to_map(Struct::getprop($this->options, "feature"));
        if ($feature_opts) {
            $items = Struct::items($feature_opts);
            if ($items) {
                foreach ($items as $item) {
                    $fname = $item[0];
                    $fopts = RailwayStationPhotosHelpers::to_map($item[1]);
                    if ($fopts && isset($fopts["active"]) && $fopts["active"] === true) {
                        ($utility->feature_add)($this->_rootctx, RailwayStationPhotosFeatures::make_feature($fname));
                    }
                }
            }
        }

        // Add extension features.
        $extend_val = Struct::getprop($this->options, "extend");
        if (is_array($extend_val)) {
            foreach ($extend_val as $f) {
                if (is_object($f) && method_exists($f, 'get_name')) {
                    ($utility->feature_add)($this->_rootctx, $f);
                }
            }
        }

        // Initialize features.
        foreach ($this->features as $f) {
            ($utility->feature_init)($this->_rootctx, $f);
        }

        ($utility->feature_hook)($this->_rootctx, "PostConstruct");
    }

    public function options_map(): array
    {
        $out = Struct::clone($this->options);
        return is_array($out) ? $out : [];
    }

    public function get_utility()
    {
        return RailwayStationPhotosUtility::copy($this->_utility);
    }

    public function get_root_ctx()
    {
        return $this->_rootctx;
    }

    public function prepare(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;
        $fetchargs = $fetchargs ?? [];

        $ctrl = RailwayStationPhotosHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "prepare",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $opts = $this->options;
        $path = Struct::getprop($fetchargs, "path") ?? "";
        $path = is_string($path) ? $path : "";
        $method_val = Struct::getprop($fetchargs, "method") ?? "GET";
        $method_val = is_string($method_val) ? $method_val : "GET";
        $params = RailwayStationPhotosHelpers::to_map(Struct::getprop($fetchargs, "params")) ?? [];
        $query = RailwayStationPhotosHelpers::to_map(Struct::getprop($fetchargs, "query")) ?? [];
        $headers = ($utility->prepare_headers)($ctx);

        $base = Struct::getprop($opts, "base") ?? "";
        $base = is_string($base) ? $base : "";
        $prefix = Struct::getprop($opts, "prefix") ?? "";
        $prefix = is_string($prefix) ? $prefix : "";
        $suffix = Struct::getprop($opts, "suffix") ?? "";
        $suffix = is_string($suffix) ? $suffix : "";

        $ctx->spec = new RailwayStationPhotosSpec([
            "base" => $base, "prefix" => $prefix, "suffix" => $suffix,
            "path" => $path, "method" => $method_val,
            "params" => $params, "query" => $query, "headers" => $headers,
            "body" => Struct::getprop($fetchargs, "body"),
            "step" => "start",
        ]);

        // Merge user-provided headers.
        $uh = Struct::getprop($fetchargs, "headers");
        if (is_array($uh)) {
            foreach ($uh as $k => $v) {
                $ctx->spec->headers[$k] = $v;
            }
        }

        [$_, $err] = ($utility->prepare_auth)($ctx);
        if ($err) {
            return ($utility->make_error)($ctx, $err);
        }

        [$fetchdef, $fd_err] = ($utility->make_fetch_def)($ctx);
        if ($fd_err) {
            return ($utility->make_error)($ctx, $fd_err);
        }
        return $fetchdef;
    }

    public function direct(array $fetchargs = []): mixed
    {
        $utility = $this->_utility;

        // direct() is the raw-HTTP escape hatch: it never throws, it returns
        // an {ok, err, ...} dict. prepare() now raises on error, so catch it
        // and surface the failure through the dict instead.
        try {
            $fetchdef = $this->prepare($fetchargs);
        } catch (\Throwable $err) {
            return ["ok" => false, "err" => $err];
        }

        $fetchargs = $fetchargs ?? [];
        $ctrl = RailwayStationPhotosHelpers::to_map(Struct::getprop($fetchargs, "ctrl")) ?? [];

        $ctx = ($utility->make_context)([
            "opname" => "direct",
            "ctrl" => $ctrl,
        ], $this->_rootctx);

        $url = $fetchdef["url"] ?? "";
        [$fetched, $fetch_err] = ($utility->fetcher)($ctx, $url, $fetchdef);

        if ($fetch_err) {
            return ["ok" => false, "err" => $fetch_err];
        }

        if ($fetched === null) {
            return [
                "ok" => false,
                "err" => $ctx->make_error("direct_no_response", "response: undefined"),
            ];
        }

        if (is_array($fetched)) {
            $status = RailwayStationPhotosHelpers::to_int(Struct::getprop($fetched, "status"));
            $headers = Struct::getprop($fetched, "headers") ?? [];

            // No-body responses (204, 304) and explicit zero content-length
            // must skip JSON parsing — calling json() on an empty body errors.
            $content_length = is_array($headers) ? ($headers["content-length"] ?? null) : null;
            $no_body = $status === 204 || $status === 304 || (string)$content_length === "0";

            $json_data = null;
            if (!$no_body) {
                $jf = Struct::getprop($fetched, "json");
                if (is_callable($jf)) {
                    try {
                        $json_data = $jf();
                    } catch (\Throwable $e) {
                        // Non-JSON body — leave data null but keep status/ok.
                        $json_data = null;
                    }
                }
            }

            return [
                "ok" => $status >= 200 && $status < 300,
                "status" => $status,
                "headers" => Struct::getprop($fetched, "headers"),
                "data" => $json_data,
            ];
        }

        return [
            "ok" => false,
            "err" => $ctx->make_error("direct_invalid", "invalid response type"),
        ];
    }


    private $_admin_inbox = null;

    // Canonical facade: $client->AdminInbox()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->admin_inbox()
    // resolves here too.
    public function AdminInbox($data = null)
    {
        require_once __DIR__ . '/entity/admin_inbox_entity.php';
        if ($data === null) {
            if ($this->_admin_inbox === null) {
                $this->_admin_inbox = new AdminInboxEntity($this, null);
            }
            return $this->_admin_inbox;
        }
        return new AdminInboxEntity($this, $data);
    }


    private $_country = null;

    // Canonical facade: $client->Country()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->country()
    // resolves here too.
    public function Country($data = null)
    {
        require_once __DIR__ . '/entity/country_entity.php';
        if ($data === null) {
            if ($this->_country === null) {
                $this->_country = new CountryEntity($this, null);
            }
            return $this->_country;
        }
        return new CountryEntity($this, $data);
    }


    private $_inbox = null;

    // Canonical facade: $client->Inbox()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->inbox()
    // resolves here too.
    public function Inbox($data = null)
    {
        require_once __DIR__ . '/entity/inbox_entity.php';
        if ($data === null) {
            if ($this->_inbox === null) {
                $this->_inbox = new InboxEntity($this, null);
            }
            return $this->_inbox;
        }
        return new InboxEntity($this, $data);
    }


    private $_inbox_count = null;

    // Canonical facade: $client->InboxCount()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->inbox_count()
    // resolves here too.
    public function InboxCount($data = null)
    {
        require_once __DIR__ . '/entity/inbox_count_entity.php';
        if ($data === null) {
            if ($this->_inbox_count === null) {
                $this->_inbox_count = new InboxCountEntity($this, null);
            }
            return $this->_inbox_count;
        }
        return new InboxCountEntity($this, $data);
    }


    private $_inbox_entry = null;

    // Canonical facade: $client->InboxEntry()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->inbox_entry()
    // resolves here too.
    public function InboxEntry($data = null)
    {
        require_once __DIR__ . '/entity/inbox_entry_entity.php';
        if ($data === null) {
            if ($this->_inbox_entry === null) {
                $this->_inbox_entry = new InboxEntryEntity($this, null);
            }
            return $this->_inbox_entry;
        }
        return new InboxEntryEntity($this, $data);
    }


    private $_inbox_state_query = null;

    // Canonical facade: $client->InboxStateQuery()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->inbox_state_query()
    // resolves here too.
    public function InboxStateQuery($data = null)
    {
        require_once __DIR__ . '/entity/inbox_state_query_entity.php';
        if ($data === null) {
            if ($this->_inbox_state_query === null) {
                $this->_inbox_state_query = new InboxStateQueryEntity($this, null);
            }
            return $this->_inbox_state_query;
        }
        return new InboxStateQueryEntity($this, $data);
    }


    private $_o_auth_token = null;

    // Canonical facade: $client->OAuthToken()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->o_auth_token()
    // resolves here too.
    public function OAuthToken($data = null)
    {
        require_once __DIR__ . '/entity/o_auth_token_entity.php';
        if ($data === null) {
            if ($this->_o_auth_token === null) {
                $this->_o_auth_token = new OAuthTokenEntity($this, null);
            }
            return $this->_o_auth_token;
        }
        return new OAuthTokenEntity($this, $data);
    }


    private $_oauth = null;

    // Canonical facade: $client->Oauth()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->oauth()
    // resolves here too.
    public function Oauth($data = null)
    {
        require_once __DIR__ . '/entity/oauth_entity.php';
        if ($data === null) {
            if ($this->_oauth === null) {
                $this->_oauth = new OauthEntity($this, null);
            }
            return $this->_oauth;
        }
        return new OauthEntity($this, $data);
    }


    private $_photo = null;

    // Canonical facade: $client->Photo()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->photo()
    // resolves here too.
    public function Photo($data = null)
    {
        require_once __DIR__ . '/entity/photo_entity.php';
        if ($data === null) {
            if ($this->_photo === null) {
                $this->_photo = new PhotoEntity($this, null);
            }
            return $this->_photo;
        }
        return new PhotoEntity($this, $data);
    }


    private $_photo_download = null;

    // Canonical facade: $client->PhotoDownload()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->photo_download()
    // resolves here too.
    public function PhotoDownload($data = null)
    {
        require_once __DIR__ . '/entity/photo_download_entity.php';
        if ($data === null) {
            if ($this->_photo_download === null) {
                $this->_photo_download = new PhotoDownloadEntity($this, null);
            }
            return $this->_photo_download;
        }
        return new PhotoDownloadEntity($this, $data);
    }


    private $_photo_station = null;

    // Canonical facade: $client->PhotoStation()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->photo_station()
    // resolves here too.
    public function PhotoStation($data = null)
    {
        require_once __DIR__ . '/entity/photo_station_entity.php';
        if ($data === null) {
            if ($this->_photo_station === null) {
                $this->_photo_station = new PhotoStationEntity($this, null);
            }
            return $this->_photo_station;
        }
        return new PhotoStationEntity($this, $data);
    }


    private $_photo_upload = null;

    // Canonical facade: $client->PhotoUpload()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->photo_upload()
    // resolves here too.
    public function PhotoUpload($data = null)
    {
        require_once __DIR__ . '/entity/photo_upload_entity.php';
        if ($data === null) {
            if ($this->_photo_upload === null) {
                $this->_photo_upload = new PhotoUploadEntity($this, null);
            }
            return $this->_photo_upload;
        }
        return new PhotoUploadEntity($this, $data);
    }


    private $_photographer = null;

    // Canonical facade: $client->Photographer()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->photographer()
    // resolves here too.
    public function Photographer($data = null)
    {
        require_once __DIR__ . '/entity/photographer_entity.php';
        if ($data === null) {
            if ($this->_photographer === null) {
                $this->_photographer = new PhotographerEntity($this, null);
            }
            return $this->_photographer;
        }
        return new PhotographerEntity($this, $data);
    }


    private $_profile = null;

    // Canonical facade: $client->Profile()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->profile()
    // resolves here too.
    public function Profile($data = null)
    {
        require_once __DIR__ . '/entity/profile_entity.php';
        if ($data === null) {
            if ($this->_profile === null) {
                $this->_profile = new ProfileEntity($this, null);
            }
            return $this->_profile;
        }
        return new ProfileEntity($this, $data);
    }


    private $_public_inbox = null;

    // Canonical facade: $client->PublicInbox()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->public_inbox()
    // resolves here too.
    public function PublicInbox($data = null)
    {
        require_once __DIR__ . '/entity/public_inbox_entity.php';
        if ($data === null) {
            if ($this->_public_inbox === null) {
                $this->_public_inbox = new PublicInboxEntity($this, null);
            }
            return $this->_public_inbox;
        }
        return new PublicInboxEntity($this, $data);
    }


    private $_stat = null;

    // Canonical facade: $client->Stat()->list() / ->load(["id" => ...]).
    // PHP method names are case-insensitive, so lowercase $client->stat()
    // resolves here too.
    public function Stat($data = null)
    {
        require_once __DIR__ . '/entity/stat_entity.php';
        if ($data === null) {
            if ($this->_stat === null) {
                $this->_stat = new StatEntity($this, null);
            }
            return $this->_stat;
        }
        return new StatEntity($this, $data);
    }



    public static function test(?array $testopts = null, ?array $sdkopts = null): self
    {
        $sdkopts = $sdkopts ?? [];
        $sdkopts = Struct::clone($sdkopts);
        $sdkopts = is_array($sdkopts) ? $sdkopts : [];

        $testopts = $testopts ?? [];
        $testopts = Struct::clone($testopts);
        $testopts = is_array($testopts) ? $testopts : [];
        $testopts["active"] = true;

        if (!isset($sdkopts["feature"])) {
            $sdkopts["feature"] = [];
        }
        $sdkopts["feature"]["test"] = $testopts;

        $sdk = new RailwayStationPhotosSDK($sdkopts);
        $sdk->mode = "test";
        return $sdk;
    }
}
